internal/repository/indexmap.go

```diff
@@ -0,0 +1,140 @@
+package repository
+
+import (
+	"encoding/binary"
+
+	"github.com/restic/restic/internal/restic"
+)
+
+// An indexMap is a chained hash table that maps blob IDs to indexEntries.
+// It allows storing multiple entries with the same key.
+//
+// A chained table has less memory overhead than a probing one (like the
+// standard map), because only a pointer array needs to be resized when the
+// table grows.
+type indexMap struct {
+	// The number of buckets is always a power of two and never zero.
+	buckets    []*indexEntry
+	numentries uint
+
+	free *indexEntry // Free list.
+}
+
+const (
+	growthFactor = 2 // Must be a power of 2.
+	maxLoad      = 4 // Max. number of entries per bucket.
```

Comment:
With `maxLoad = 4` we have minimum 2 and maximum 4 entries per bucket. For each bucket we have to store a pointer of 8 bytes. Hence this is an overhead of 2 to 4 bytes per Entry. I would advise to choose a slightly higher number, e.g. 6 or 8. This will decrease Lookup speed a bit but may easily save another byte per entry. IMO a much higher Lookup time can be afforded once we merge indexes into one (actually Lookup loops over maybe hundreds of index files).

This proposed change however can easily be postponed to the following merge PR :wink:

Reply:
Then I say we postpone. We still need good lookup performance for #2787 in the meantime.